File: packages/core/lib/src/passenger.dart
import 'dart:math' as math;

import 'api.dart';
import 'arabic.dart';
import 'corridor.dart';
import 'geo.dart';
import 'models.dart';
import 'transport.dart';

/// A place she can ask for, and every line that would take her there.
class DestinationOption {
  const DestinationOption({
    required this.id,
    required this.nameAr,
    required this.routeIds,
  });

  final String id;
  final String nameAr;
  final List<String> routeIds;
}

/// A bus she could catch, with the arithmetic done on her phone.
class BusSighting {
  const BusSighting({
    required this.pseudonym,
    required this.routeId,
    required this.dir,
    required this.gapM,
    required this.etaSeconds,
  });

  final String pseudonym;
  final String routeId;
  final int dir;

  /// How far the bus still has to travel before it reaches her.
  final double gapM;

  /// Null when the bus is stopped, because a number would be a guess.
  final double? etaSeconds;

  /// "بعد ٦ دقايق"
  String get etaText {
    if (etaSeconds == null) return 'الباص واقف';
    final minutes = (etaSeconds! / 60).round();
    if (minutes <= 1) return 'وصل تقريباً';
    return 'بعد $minutes دقايق';
  }
}

/// Which line, and which way, would take her where she is going.
class RideOption {
  const RideOption({
    required this.route,
    required this.dir,
    required this.remainingM,
    required this.zoneSeq,
    required this.destinationId,
  });

  final Route route;
  final Direction dir;

  /// Her own distance from that direction's endpoint, computed locally.
  final double remainingM;
  final int zoneSeq;
  final String destinationId;
}

/// Where a waiting passenger is in the one interaction she has.
enum WaitState { idle, waiting, boarded, cancelled, expired }

/// A passenger's whole journey through the app.
///
/// She has no account and sends no coordinate: her phone holds the network,
/// works out which lines pass her and how far along she stands, and sends only
/// that.
class PassengerController {
  PassengerController({
    required MowasalatApi api,
    required List<Route> network,
    required double remainingBucketM,
    int Function()? now,
  })  : _api = api,
        _network = network,
        _bucketM = remainingBucketM,
        _now = now ?? (() => DateTime.now().millisecondsSinceEpoch);

  static const requestLifetime = Duration(minutes: 20);

  final MowasalatApi _api;
  final List<Route> _network;
  final double _bucketM;
  final int Function() _now;

  final Map<String, CorridorMatcher> _matchers = <String, CorridorMatcher>{};

  WaitState _state = WaitState.idle;
  String? _pseudonym;
  RideOption? _riding;
  int _requestedAtMs = 0;

  WaitState get state => _state;
  String? get requestPseudonym => _pseudonym;
  RideOption? get chosenRide => _riding;

  CorridorMatcher _matcherFor(Route route) =>
      _matchers.putIfAbsent(route.id, () => CorridorMatcher(route));

  /// Everywhere the network can take her, for the وين رايح؟ tiles and search.
  List<DestinationOption> destinations({String? query}) {
    final byId = <String, List<String>>{};
    final names = <String, String>{};

    for (final route in _network) {
      for (final served in route.servedDestinations) {
        names[served.id] = served.nameAr;
        byId.putIfAbsent(served.id, () => <String>[]).add(route.id);
      }
    }

    final options = byId.entries
        .map((e) => DestinationOption(id: e.key, nameAr: names[e.key]!, routeIds: e.value))
        .toList()
      ..sort((a, b) => a.nameAr.compareTo(b.nameAr));

    if (query == null || query.trim().isEmpty) return options;
    return options.where((o) => matchesPlace(query, o.nameAr)).toList(growable: false);
  }

  /// Which lines would actually pick her up, given where she is standing.
  ///
  /// A line only counts when her destination is still *ahead* of her on it —
  /// standing past the village she wants is not a ride, it is a walk back.
  List<RideOption> ridesFor({required String destinationId, required LatLng position}) {
    final options = <RideOption>[];

    for (final route in _network) {
      final serves = route.servedDestinations.where((d) => d.id == destinationId);
      if (serves.isEmpty) continue;

      final matcher = _matcherFor(route);
      final destinationZone = serves.first.zoneSeq;
      final destinationCentre = _zoneCentre(route, destinationZone);
      if (destinationCentre == null) continue;

      for (final dir in <Direction>[0, 1]) {
        final mine = matcher.match(position, dir);
        if (!mine.inside) continue;

        // Her destination is ahead when it has less left to run than she does.
        final theirs = matcher.match(destinationCentre, dir);
        if (theirs.remainingM >= mine.remainingM) continue;

        options.add(RideOption(
          route: route,
          dir: dir,
          remainingM: mine.remainingM,
          zoneSeq: mine.zoneSeq,
          destinationId: destinationId,
        ));
      }
    }

    // The shortest ride first: the line whose endpoint she is nearest.
    options.sort((a, b) => a.remainingM.compareTo(b.remainingM));
    return options;
  }

  LatLng? _zoneCentre(Route route, int seq) {
    for (final z in route.corridor.zones) {
      if (z.seq == seq) return z.centre;
    }
    return null;
  }

  /// Asks which buses are running, and keeps the ticket that lets her watch.
  Future<({List<BusSighting> buses, String streamTicket})> busesFor(RideOption ride) async {
    final result = await _api.findBuses(
      routeId: ride.route.id,
      dir: ride.dir,
      remainingM: bucketRemaining(ride.remainingM, _bucketM),
      zoneSeq: ride.zoneSeq,
    );

    return (
      buses: result.buses
          .map((b) => BusSighting(
                pseudonym: b['pseudonym'] as String,
                routeId: ride.route.id,
                dir: ride.dir,
                gapM: (b['gapM'] as num).toDouble(),
                etaSeconds: (b['etaSeconds'] as num?)?.toDouble(),
              ))
          .toList(growable: false),
      streamTicket: result.streamTicket,
    );
  }

  /// Turns a stream snapshot into sightings.
  ///
  /// The stream sends bus scalars, not answers: the gap and the ETA are worked
  /// out here, which is also why she never has to send her position to watch.
  List<BusSighting> sightingsFrom(
    List<Map<String, dynamic>> buses,
    RideOption ride, {
    double windowM = 15000,
  }) {
    final sightings = <BusSighting>[];

    for (final bus in buses) {
      final busRemaining = (bus['remainingM'] as num).toDouble();
      final speedKph = (bus['speedKph'] as num?)?.toDouble() ?? 0;

      // A bus passes her only if it still has further to go than she does.
      final gap = busRemaining - ride.remainingM;
      if (gap <= 0 || gap > windowM) continue;

      sightings.add(BusSighting(
        pseudonym: bus['pseudonym'] as String,
        routeId: ride.route.id,
        dir: ride.dir,
        gapM: gap,
        etaSeconds: speedKph < 3 ? null : (gap / (speedKph * 1000)) * 3600,
      ));
    }

    sightings.sort((a, b) => a.gapM.compareTo(b.gapM));
    return sightings;
  }

  /// أنا مستني هون.
  Future<String> requestRide(RideOption ride) async {
    if (_state == WaitState.waiting) throw StateError('already waiting');

    final pseudonym = await _api.requestRide(RideRequestPayload(
      routeId: ride.route.id,
      dir: ride.dir,
      destinationId: ride.destinationId,
      remainingM: bucketRemaining(ride.remainingM, _bucketM),
      zoneSeq: ride.zoneSeq,
    ));

    _pseudonym = pseudonym;
    _riding = ride;
    _requestedAtMs = _now();
    _state = WaitState.waiting;
    return pseudonym;
  }

  /// True once the backend would have dropped it anyway.
  bool get hasExpired =>
      _state == WaitState.waiting &&
      _now() - _requestedAtMs >= requestLifetime.inMilliseconds;

  Duration get timeLeft {
    if (_state != WaitState.waiting) return Duration.zero;
    final left = requestLifetime.inMilliseconds - (_now() - _requestedAtMs);
    return Duration(milliseconds: math.max(0, left));
  }

  /// Moves to expired once the window has passed, so the screen stops
  /// promising a bus that will never be told about her.
  WaitState refresh() {
    if (hasExpired) _state = WaitState.expired;
    return _state;
  }

  /// إلغاء — removed at once, everywhere.
  Future<void> cancel() async {
    final pseudonym = _pseudonym;
    if (pseudonym == null) return;
    try {
      await _api.cancelRequest(pseudonym);
    } on ApiException {
      // Already gone server-side; the local state still has to be cleared.
    }
    _clear(WaitState.cancelled);
  }

  /// ركبت.
  Future<void> boarded() async {
    final pseudonym = _pseudonym;
    final ride = _riding;
    if (pseudonym == null || ride == null) return;
    try {
      await _api.boarded(pseudonym: pseudonym, routeId: ride.route.id, dir: ride.dir);
    } on ApiException {
      // The ride happened either way.
    }
    _clear(WaitState.boarded);
  }

  void _clear(WaitState finalState) {
    _pseudonym = null;
    _riding = null;
    _requestedAtMs = 0;
    _state = finalState;
  }

  /// Back to وين رايح؟ for the next journey.
  void reset() => _clear(WaitState.idle);
}

File: apps/web-passenger/src/passenger.ts
/**
 * A passenger's whole journey through the app.
 *
 * She has no account and sends no coordinate: the page holds the network,
 * works out which lines pass her and how far along she stands, and sends only
 * that. A port of packages/core/lib/src/passenger.dart — the two clients must
 * agree, so the arithmetic is the same arithmetic.
 */

import {
  indexCorridor,
  place,
  remainingM as remainingAlong,
  bucketRemaining,
  type IndexedCorridor,
} from "../../../packages/corridor/src/index.ts";
import type { Direction, LatLng, Route } from "../../../packages/corridor/src/types.ts";
import { matchesPlace } from "./arabic.ts";
import type { Api, BusScalar } from "./api.ts";

/** A place she can ask for, and every line that would take her there. */
export type DestinationOption = {
  id: string;
  nameAr: string;
  routeIds: string[];
};

/** Which line, and which way, would take her where she is going. */
export type RideOption = {
  route: Route;
  dir: Direction;
  /** Her own distance from that direction's endpoint, computed locally. */
  remainingM: number;
  zoneSeq: number;
  destinationId: string;
};

/** A bus she could catch, with the arithmetic done on her phone. */
export type BusSighting = {
  pseudonym: string;
  routeId: string;
  dir: Direction;
  /** How far the bus still has to travel before it reaches her. */
  gapM: number;
  /** Null when the bus is stopped, because a number would be a guess. */
  etaSeconds: number | null;
};

export type WaitState = "idle" | "waiting" | "boarded" | "cancelled" | "expired";

export const REQUEST_LIFETIME_MS = 20 * 60 * 1000;

/** Below this a bus is stopped and an ETA would be invented. */
const MIN_SPEED_KPH = 3;

/** Buses further back than this are not a ride she is waiting for. */
const WINDOW_M = 15_000;

export function etaText(sighting: BusSighting): string {
  if (sighting.etaSeconds === null) return "الباص واقف";
  const minutes = Math.round(sighting.etaSeconds / 60);
  if (minutes <= 1) return "وصل تقريباً";
  return `بعد ${minutes} دقايق`;
}

export class PassengerController {
  private readonly matchers = new Map<string, IndexedCorridor>();

  private _state: WaitState = "idle";
  private _pseudonym: string | null = null;
  private _riding: RideOption | null = null;
  private requestedAtMs = 0;

  private readonly api: Api;
  private readonly network: Route[];
  private readonly bucketM: number;
  private readonly now: () => number;

  constructor(
    api: Api,
    network: Route[],
    bucketM: number,
    now: () => number = () => Date.now(),
  ) {
    // Without a band there is nothing to blur her position to. Failing here
    // means a misconfigured server shows an error rather than an app that
    // looks like it is protecting her and is not.
    if (!Number.isFinite(bucketM) || bucketM <= 0) {
      throw new Error("the country's remainingBucketM is missing or invalid");
    }

    this.api = api;
    this.network = network;
    this.bucketM = bucketM;
    this.now = now;
  }

  get state(): WaitState {
    return this._state;
  }
  get requestPseudonym(): string | null {
    return this._pseudonym;
  }
  get chosenRide(): RideOption | null {
    return this._riding;
  }

  private indexed(route: Route): IndexedCorridor {
    let ic = this.matchers.get(route.id);
    if (!ic) {
      ic = indexCorridor(route.corridor);
      this.matchers.set(route.id, ic);
    }
    return ic;
  }

  private fix(route: Route, p: LatLng, dir: Direction) {
    const ic = this.indexed(route);
    const placement = place(p, ic);
    return {
      inside: placement.inside,
      zoneSeq: placement.zoneSeq,
      remainingM: remainingAlong(placement, ic, dir),
    };
  }

  /** Everywhere the network can take her, for the وين رايح؟ tiles and search. */
  destinations(query?: string): DestinationOption[] {
    const byId = new Map<string, string[]>();
    const names = new Map<string, string>();

    for (const route of this.network) {
      for (const served of route.servedDestinations) {
        names.set(served.id, served.nameAr);
        const list = byId.get(served.id) ?? [];
        list.push(route.id);
        byId.set(served.id, list);
      }
    }

    const options = [...byId.entries()]
      .map(([id, routeIds]) => ({ id, nameAr: names.get(id)!, routeIds }))
      .sort((a, b) => a.nameAr.localeCompare(b.nameAr, "ar"));

    if (!query || !query.trim()) return options;
    return options.filter((o) => matchesPlace(query, o.nameAr));
  }

  /**
   * Which lines would actually pick her up, given where she is standing.
   *
   * A line only counts when her destination is still *ahead* of her on it —
   * standing past the village she wants is not a ride, it is a walk back.
   */
  ridesFor(destinationId: string, position: LatLng): RideOption[] {
    const options: RideOption[] = [];

    for (const route of this.network) {
      const served = route.servedDestinations.find((d) => d.id === destinationId);
      if (!served) continue;

      const destinationCentre = route.corridor.zones.find((z) => z.seq === served.zoneSeq)?.centre;
      if (!destinationCentre) continue;

      for (const dir of [0, 1] as Direction[]) {
        const mine = this.fix(route, position, dir);
        if (!mine.inside) continue;

        // Her destination is ahead when it has less left to run than she does.
        const theirs = this.fix(route, destinationCentre, dir);
        if (theirs.remainingM >= mine.remainingM) continue;

        options.push({
          route,
          dir,
          remainingM: mine.remainingM,
          zoneSeq: mine.zoneSeq,
          destinationId,
        });
      }
    }

    // The shortest ride first: the line whose endpoint she is nearest.
    return options.sort((a, b) => a.remainingM - b.remainingM);
  }

  /** Asks which buses are running, and keeps the ticket that lets her watch. */
  async busesFor(ride: RideOption): Promise<{ buses: BusSighting[]; streamTicket: string }> {
    const result = await this.api.findBuses({
      routeId: ride.route.id,
      dir: ride.dir,
      remainingM: bucketRemaining(ride.remainingM, this.bucketM),
      zoneSeq: ride.zoneSeq,
    });

    return {
      buses: result.buses.map((b) => ({
        pseudonym: b.pseudonym,
        routeId: ride.route.id,
        dir: ride.dir,
        gapM: b.gapM ?? 0,
        etaSeconds: b.etaSeconds ?? null,
      })),
      streamTicket: result.streamTicket,
    };
  }

  /**
   * Turns a stream snapshot into sightings.
   *
   * The stream sends bus scalars, not answers: the gap and the ETA are worked
   * out here, which is also why she never has to send her position to watch.
   */
  sightingsFrom(buses: BusScalar[], ride: RideOption, windowM = WINDOW_M): BusSighting[] {
    const sightings: BusSighting[] = [];

    for (const bus of buses) {
      const speedKph = bus.speedKph ?? 0;

      // A bus passes her only if it still has further to go than she does.
      const gap = bus.remainingM - ride.remainingM;
      if (gap <= 0 || gap > windowM) continue;

      sightings.push({
        pseudonym: bus.pseudonym,
        routeId: ride.route.id,
        dir: ride.dir,
        gapM: gap,
        etaSeconds: speedKph < MIN_SPEED_KPH ? null : (gap / (speedKph * 1000)) * 3600,
      });
    }

    return sightings.sort((a, b) => a.gapM - b.gapM);
  }

  /** أنا مستني هون. */
  async requestRide(ride: RideOption): Promise<string> {
    if (this._state === "waiting") throw new Error("already waiting");

    const { pseudonym } = await this.api.requestRide({
      routeId: ride.route.id,
      dir: ride.dir,
      destinationId: ride.destinationId,
      remainingM: bucketRemaining(ride.remainingM, this.bucketM),
      zoneSeq: ride.zoneSeq,
    });

    this._pseudonym = pseudonym;
    this._riding = ride;
    this.requestedAtMs = this.now();
    this._state = "waiting";
    return pseudonym;
  }

  /** True once the backend would have dropped it anyway. */
  get hasExpired(): boolean {
    return this._state === "waiting" && this.now() - this.requestedAtMs >= REQUEST_LIFETIME_MS;
  }

  get timeLeftMs(): number {
    if (this._state !== "waiting") return 0;
    return Math.max(0, REQUEST_LIFETIME_MS - (this.now() - this.requestedAtMs));
  }

  /**
   * Moves to expired once the window has passed, so the screen stops promising
   * a bus that will never be told about her.
   */
  refresh(): WaitState {
    if (this.hasExpired) this._state = "expired";
    return this._state;
  }

  /** إلغاء — removed at once, everywhere. */
  async cancel(): Promise<void> {
    const pseudonym = this._pseudonym;
    if (!pseudonym) return;
    try {
      await this.api.cancelRequest(pseudonym);
    } catch {
      // Already gone server-side; the local state still has to be cleared.
    }
    this.clear("cancelled");
  }

  /** ركبت. */
  async boarded(): Promise<void> {
    const pseudonym = this._pseudonym;
    const ride = this._riding;
    if (!pseudonym || !ride) return;
    try {
      await this.api.boarded({ pseudonym, routeId: ride.route.id, dir: ride.dir });
    } catch {
      // The ride happened either way.
    }
    this.clear("boarded");
  }

  /** Back to وين رايح؟ for the next journey. */
  reset(): void {
    this.clear("idle");
  }

  private clear(finalState: WaitState): void {
    this._pseudonym = null;
    this._riding = null;
    this.requestedAtMs = 0;
    this._state = finalState;
  }
}
